content/blog/2023/2024-01-27-deserialize-with-jsoncreator.md

```diff
@@ -0,0 +1,450 @@
+````java
+
+    @Test
+    public void deserializeImmutableObjects() throws JsonProcessingException {
+        String userData = 
+            objectMapper.writeValueAsString(MockedUsersUtility.getMockedUserData());
+        System.out.println("USER: " + userData);
+        UserData user = objectMapper.readValue(userData, UserData.class);
+        assertNotNull(user);
+    }
+    
+````
+In the above example the `writeValueAsString()` serializes the `UserData` object to a String.
+The `readValue()` method is responsible for deserializing the String Object back to the `UserData` object.
+When we run the test above, we see this error:
+````text
+com.fasterxml.jackson.databind.exc.InvalidDefinitionException: 
+Cannot construct instance of `com.reflectoring.userdetails.model.UserData` 
+(no Creators, like default constructor, exist): 
+cannot deserialize from Object value (no delegate- or property-based Creator)
+ at [Source: (String)"{"id":100,"firstName":"Ranjani","lastName":"Harish",
+ createdDate":"2024-01-16"}"; line: 1, column: 2]
+````
+**The error message explicitly states that the test could not run successfully as there was an error during deserialization**.
+This is because the `ObjectMapper` class, by default, looks for a no-arg constructor to set the values. **Since our Java class is immutable, it has neither a no-arg constructor
+nor setter methods to set values**. Also, we see that the `UserData` class has multiple constructors. Therefore, we would need a way to 
+instruct the `ObjectMapper` class to use the correct constructor for deserialization.
+Let's modify our code to use the `@JsonCreator` annotation as :
+````java
+    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+    public UserData(@JsonProperty("id") long id, 
+                    @JsonProperty("firstName") String firstName,
+                    @JsonProperty("lastName") String lastName) {
+        this.id = id;
+        this.firstName = firstName;
+        this.lastName = lastName;
+        this.createdDate = LocalDate.now();
+    }
+````
+Now, when we run our test again, we can see that the deserialization is successful.
+Let's look at the additional annotation we've added to get this working.
+- By applying the `@JsonCreator` annotation to the constructor, the Jackson deserializer knows which constructor needs to be used.
+- `JsonCreator.Mode.PROPERTIES` indicates that the creator should match the incoming object with the constructor arguments. This is the most commonly used JsonCreator Mode.
+- We annotate all the constructor arguments with `@JsonProperty` for the creator to map the arguments.
+
+### Understanding All Available JsonCreator Modes
+
+We can pass one of these four values as parameters to this annotation:
+- **JsonCreator.Mode.PROPERTIES** : This is the most commonly used mode where every constructor/factory argument is annotated with either `@JsonProperty` to indicate the name of the property to bind to.
```

Comment:
This sentence is incomplete. It should probably be something like "with either `@JsonProperty` to indicate the name of the property to bind to, or ....".